File: 238. Product of Array Except Self/238. Product of Array Except Self.c
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* productExceptSelf(int* nums, int numsSize, int* returnSize) {
    int* prods = (int *)malloc(sizeof(int) * numsSize);
    int i = 0, prod;
    // calculate the products by two sides
    for (i = 0, prod = 1; i < numsSize; i++) {
        prods[i] = prod;
        prod *= nums[i];
    }
    for (i = numsSize - 1, prod = 1; i >= 0; i--) {
        prods[i] *= prod;
        prod *= nums[i];
    }
    *returnSize = numsSize;
    return prods;
}

File: 238. Product of Array Except Self/238. Product of Array Except Self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    var l = nums.length
    var prods = new Array(l)
    for (var i = 0, prod = 1; i < l; i++) {
        prods[i] = prod
        prod *= nums[i]
    }
    for (var i = l - 1, prod = 1; i >= 0; i--) {
        prods[i] *= prod
        prod *= nums[i]
    }
    return prods
};
